utils/estimated_histogram.hh

```diff
@@ -47,13 +47,302 @@
+ * All values higher than MAX will be included the last bucket that serves as the
+ * infinity bucket.
+ *
+ * buckets are distributed as pseudo floating point:
+ * The range [MIN, MAX) is split into log2 ranges.
+ *     ranges = log2(max/min)
+ * Each of that ranges is split according to the number of buckets:
+ *     resolution = (NUM_BUCKETS - 1)/ranges
+ *
+ * For example, if the MIN value is 128, the MAX is 1024 and the number of buckets is 13:
+ *
+ * Anything below 128 will be in the bucket 0, anything above 1024 will be in bucket 13.
+ *
+ * the range [128, 1024) will be split into log2(1024/128) = 3:
+ * 128, 256, 512, 1024
+ *
+ * Each range is split into 12/3 = 4.
+ * 128            | 256            | 512            | 1024
+ * 128 160 192 224| 256 320 384 448| 512 640 768 896|
+ *
+ *
+ * Calculating the bucket limit of bucket i:
+ * The range: 2^(i/4)* min
+ * The sub range: i%4 * range/4
+ *
+ * How to find a bucket index for a value.
+ * The bucket index consist of two part:
+ * higher bits are based on log2(value/min)
+ *
+ * lower bits are based on the high 2 MSB (ignoring the leading 1).
+
+ * for example: 330 (101001010)
+ * higher bits: log2(330/128) = 1
+ * low bit MSB: 330 = 01 (the lower two bits out of the upper 3)
+ * So the index: 101 = 5
+ *
+ *
+ * About the min/max and number of buckets.
+ * ========================================
+ *
+ * For MIN and MAX choose numbers that are a power of 2.
+ * The number of buckets will determine the resolution that should also be a power of 2
+ * So the total number of bucket should be log2(MAX/MIN) * Resolution + 1
+ *
+ * Limitation: You must set the MIN value to be higher then the resolution.
+ * For example, for a 2 bits resolution MIN should be 2^2 = 4 or higher.
+ *
+ */
+template<uint64_t Min, uint64_t Max, size_t NumBuckets>
+requires (Min > 0 && Min < Max)
+class approx_exponential_histogram {
+    std::array<uint64_t, NumBuckets> _buckets;
+public:
+
+    static constexpr unsigned RANGES = log2floor(Max/Min);
+    static constexpr unsigned RESOLUTION = (NumBuckets - 1)/RANGES;
+    static constexpr unsigned RESOLUTION_BITS = log2floor(RESOLUTION);
+    static constexpr unsigned BASESHIFT = (Min == 0) ? 0 : log2floor(Min);
+    static constexpr uint64_t LOWER_BITS_MASK = (1 << RESOLUTION_BITS) - 1;
+    static_assert(BASESHIFT >= RESOLUTION);
+
+    approx_exponential_histogram() {
+        clear();
+    }
+
+    /*!
+     * \brief Returns the bucket lower limit given the bucket id.
+     * The first and last bucket will always return the MIN and MAX respectively.
+     *
+     */
+    uint64_t get_bucket_lower_limit(uint16_t bucket_id) const {
+        if (bucket_id == 0) {
+            return Min;
+        }
+        if (bucket_id == NumBuckets - 1) {
+            return Max;
+        }
+        int16_t range_id = (bucket_id >> RESOLUTION_BITS);
+        return (1 << (range_id + BASESHIFT)) +  ((bucket_id & LOWER_BITS_MASK) << (range_id + BASESHIFT - RESOLUTION_BITS));
```

Comment:
it was because I didn't relly on Min being a power of two, once I'll enforce that, I can do the change

Reply:
@amnonh also, I think we should protect against overflow of `range_id` by capping `NumBuckets`.

Also, since we return uint64_t, shouldn't we calculate `(uint64_t(1) << (range_id + BASESHIFT))`?
(the rest of the expression should be in `uint64_t` implicitly because of `& LOWER_BITS_MASK`, but we can make it more explicit so it will be more rebost to changes.